src/s_inter.c

```diff
@@ -138,6 +203,514 @@ void sys_set_temppath(void);
+    ret = rb_read_from_buffer(rb, buf, actualLength);
+    if(ret)
+    {
+        errno = EPERM;
+        error("Error while reading from ring_buffer in rbskt_recv");
+        return -1;
+    }
+    if(msglen > buflen) {
+        if(rbskt->rs_preserve_boundaries) {
+            printf("buflen: %zu, msglen: %zd, actualLength: %d\n", buflen, msglen, actualLength);
+            printf("warning: incoming packet truncated from %d to %d bytes.", ret, INBUFSIZE);
+            // drain buffer
+            char dest;
+            while(1 == rb_read_from_buffer(rb, &dest, 1))
+                    ;
+        }
+    }
+    return actualLength;
+}
+
+// if this is called with (1), then the audio thread will not manage the I/O
+// and the user will be responsible to regularly call sys_doio()
+void sys_dontmanageio(int status)
+{
+    pd_this->pd_inter->i_dontmanageio = status;
+}
+
+// register a callback to delete an object when a send fd fails in the IO
+// thread. This allows to report errors asynchronously
+void sys_addsendfdrmfn(int fd, t_fdsendrmfn fn, void* ptr)
+{
+    int nfd, size;
+    t_fdsend *fs;
+    if (!pd_this->pd_inter->i_fdsend)
+    {
+        /* create an empty FD list */
+        pd_this->pd_inter->i_fdsend = (t_fdsend *)t_getbytes(0);
+        pd_this->pd_inter->i_nfdsend= 0;
+    }
+    if(rbrmfdsend_init())
+        return;
+    nfd = pd_this->pd_inter->i_nfdsend;
+    size = nfd * sizeof(t_fdsend);
+    pd_this->pd_inter->i_fdsend = (t_fdsend *)t_resizebytes(
+        pd_this->pd_inter->i_fdsend, size, size + sizeof(t_fdsend));
+    fs = pd_this->pd_inter->i_fdsend + nfd;
+    fs->fds_fd = fd;
+    fs->fds_fn = fn;
+    fs->fds_ptr = ptr;
+    pd_this->pd_inter->i_nfdsend = nfd + 1;
+}
+
+// a partial replacement for the functionalities of sys_domicrosleep() when
+// enabling THREADED_IO. Namely:
+// - check for ready file descriptors, and for each ready one:
+//  - if it has a RT-friendly callback, retrieve the new data from the
+//  socket and place it in a ring buffer. sys_domicrosleep() will then call the
+//  callback which will retrieve the data from the buffer
+//  - otherwise, mark it so that the callback will be called from
+//  sys_domicrosleep()) at the first occasion
+//
+// As a consequence of this:
+//  - for callbacks that are RT-friendly, sys_domicrosleep() does not need to select()
+//  - for callbacks that are not RT-friendly, sys_domicrosleep() doesn't have to select()
+//  at every call, but only when data is actually available.
+static void poll_fds()
+{
+    const unsigned int bufsize = pd_this->pd_inter->i_iothreadbufsize;
+    ssize_t ret;
+    struct timeval timout = {0}; // making this timeout longer would be more efficient (by avoiding spurious wakeups), but it would needlessly postpone writes
+    int pollem = 1;
+    int i;
+    t_fdpoll *fp;
+    if (pollem && pd_this->pd_inter->i_nfdpoll)
+    {
+        fd_set readset;
+        fdp_select(&readset, timout, kFdpManagerIoThread);
```

Comment:
Better call `fdp_select` in a loop until there are no sockets available for reading. Otherwise you would only receive a single pending UDP packet per socket and go to sleep. This can drastically limit throughput and the OS socket buffer can overflow, leading to packet loss.

BTW, Pd currently calls `sys_microsleep` in a loop in the scheduler and only goes to sleep if there's nothing to receive (or stops if audio needs to be computed).

Reply:
> Better call fdp_select in a loop until there are no sockets available for reading. Otherwise you would only receive a single pending UDP packet per socket and go to sleep. 

oh right! This got me earlier on in `sys_domicrosleep()`(where there is now a `while()`). I'll fix it also here.

> BTW, Pd currently calls sys_microsleep in a loop in the scheduler and only goes to sleep if there's nothing to receive (or stops if audio needs to be computed).

So I could return `donesomething` from `sys_doio()` so the caller would know whether to sleep or not. (this in addition to adding the  `while()` inside `poll_fds()` as well.)

Reply:
> So I could return donesomething from sys_doio() so the caller would know whether to sleep or not.

That's actually a good idea!

> this in addition to adding the while() inside poll_fds() as well.

This would be redundant then. The while() should be in `poll_thread_loop`, right?

Reply:
Yes, but I am thinking about a poor `libpd()` user who calls `sys_doio()` in their own thread but didn't read the fine print, ignores the return value and keeps sleeping even if more data is available.

Reply:
If `sys_doio` would always keep receiving while data is available, what would be the point of the return value then?

> Yes, but I am thinking about a poor libpd() user who calls sys_doio() in their own thread

If a libpd user goes so far as managing their own I/O thread, they should know what they are doing... 

> but didn't read the fine print

Then make it large print ;-)

Reply:
da5749b8

> Then make it large print ;-)

I find the Pd public API so largely undocumented that I feel almost embarrassed about adding a comment above a function. Done anyhow!

Reply:
:-D Well, someone has to start and set a good example!